BRP improve json schema support

Leinnan opened the pull request:
# Objective

Currently implemented JSON Schema export is often producing invalid schemas for types resulting in that json that would be a valid value for schema generated for type cannot be deserialized to that type by Bevy deserializer. 

## Solution

This PR tries to reimagine how the schema is constructed. It follows Draft 2020 of JSON schema, as definied here: https://json-schema.org/draft/2020-12/schema 
It is possible to use schemas generated by new system as validation data for assets in various editors. Once there would be a support for getting information about supported extensions with reflection system I could write example. Related task: #19399 
Main additions and change to previous behaviour would be:
- `f32` generates `{"type": "number", "shortPath": "f32", "typePath": "f32"}` instead of `{"$ref": "#/$defs/f32"}`
- primitive types also are getting proper support for minimum and maximum values- from type itself and also from reflection range data, if provided.
- `CustomInternalSchemaData` enables providing custom schemas for reflected types
- Particularly useful for types like `Vec3` that serialize as 3-element arrays despite being structs with 3 fields
- `ExternalSchemaSource` trait allows referencing external schema sources
- Automatic registration for math types (Vec2, Vec3, Vec4, quaternions, matrices) to serialize as arrays
- Types with `ReflectDefault` implementation automatically include a `default` field in their schema
- Type and field descriptions are read from documentation attributes and included in schemas
- Range attributes (`#[reflect(@min..=max)]`) are properly handled
- `Option<T>` is now treated as a non required field with possible values of `T` or `null`. This approach ensures compatibility with JSON Schema validation while maintaining type safety. More specific info about whether field is required should be provided by bevy reflection system later on. Related task: #19557 
- New field `rustFieldsInfo`- It is provided when type is serialized as array, but the type is not an array. It is done to provide additional information about the fields in the schema.

Example schema:
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "shortPath": "DAffine3",
  "typePath": "glam::DAffine3",
  "modulePath": "glam",
  "crateName": "glam",
  "reflectTypeData": [
    "Deserialize",
    "Serialize",
    "Default"
  ],
  "kind": "Struct",
  "rustFieldsInfo": {
    "translation": {
      "$ref": "#/$defs/glam%3A%3ADVec3",
      "title": "translation",
      "typePath": "glam::DVec3",
      "kind": "Struct"
    },
    "matrix3": {
      "$ref": "#/$defs/glam%3A%3ADMat3",
      "title": "matrix3",
      "typePath": "glam::DMat3",
      "kind": "Struct"
    }
  },
  "type": "array",
  "prefixItems": [
    {
      "title": "matrix3.x_axis.x",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.x_axis.y",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.x_axis.z",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.y_axis.x",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.y_axis.y",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.y_axis.z",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.z_axis.x",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.z_axis.y",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "matrix3.z_axis.z",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "translation.x",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "translation.y",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    },
    {
      "title": "translation.z",
      "typePath": "f64",
      "kind": "Value",
      "type": "number"
    }
  ],
  "minItems": 12,
  "maxItems": 12,
  "$defs": {
    "glam::DMat3": {
      "shortPath": "DMat3",
      "typePath": "glam::DMat3",
      "modulePath": "glam",
      "crateName": "glam",
      "reflectTypeData": [
        "Deserialize",
        "Serialize",
        "Default"
      ],
      "kind": "Struct",
      "rustFieldsInfo": {
        "x_axis": {
          "$ref": "#/$defs/glam%3A%3ADVec3",
          "title": "x_axis",
          "typePath": "glam::DVec3",
          "kind": "Struct"
        },
        "y_axis": {
          "$ref": "#/$defs/glam%3A%3ADVec3",
          "title": "y_axis",
          "typePath": "glam::DVec3",
          "kind": "Struct"
        },
        "z_axis": {
          "$ref": "#/$defs/glam%3A%3ADVec3",
          "title": "z_axis",
          "typePath": "glam::DVec3",
          "kind": "Struct"
        }
      },
      "type": "array",
      "prefixItems": [
        {
          "title": "x_axis.x",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "x_axis.y",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "x_axis.z",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "y_axis.x",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "y_axis.y",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "y_axis.z",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "z_axis.x",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "z_axis.y",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "z_axis.z",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        }
      ],
      "minItems": 9,
      "maxItems": 9
    },
    "glam::DVec3": {
      "shortPath": "DVec3",
      "typePath": "glam::DVec3",
      "modulePath": "glam",
      "crateName": "glam",
      "reflectTypeData": [
        "Deserialize",
        "Serialize",
        "Default"
      ],
      "kind": "Struct",
      "rustFieldsInfo": {
        "x": {
          "title": "x",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        "y": {
          "title": "y",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        "z": {
          "title": "z",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        }
      },
      "type": "array",
      "prefixItems": [
        {
          "title": "x",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "y",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        },
        {
          "title": "z",
          "typePath": "f64",
          "kind": "Value",
          "type": "number"
        }
      ],
      "minItems": 3,
      "maxItems": 3
    }
  }
}
```

## Testing
- New tests are added
- I use this code for generating schemas for my assets with configurations for Zed and VSCode editor

Leinnan committed:
Schema Types Metadata

Leinnan committed:
Cleanup, init resource

Leinnan committed:
Extra test for schema

Leinnan committed:
Update mod.rs

Co-authored-by: Gino Valente <49806985+MrGVSV@users.noreply.github.com>

Leinnan committed:
Update mod.rs

Co-authored-by: Gino Valente <49806985+MrGVSV@users.noreply.github.com>

Leinnan committed:
Apply feedback

Leinnan committed:
Merge branch 'main' into schema-types-metadata

Leinnan committed:
Format fix for toml file

Leinnan committed:
Format fix for toml file

Leinnan committed:
Merge branch 'bevyengine:main' into schema-types-metadata

Leinnan committed:
JSON Schema reforge

Leinnan committed:
Merge branch 'bevyengine:main' into proper-json-schema

Leinnan committed:
Reflect info and stuff

Leinnan committed:
Default values for schemas

Leinnan committed:
ReflectJsonSchema for custom schemas

Leinnan committed:
Merge remote-tracking branch 'origin/main' into proper-json-schema

# Conflicts:
#	crates/bevy_remote/Cargo.toml
#	crates/bevy_remote/src/builtin_methods.rs
#	crates/bevy_remote/src/schemas/json_schema.rs
#	crates/bevy_remote/src/schemas/mod.rs

Leinnan committed:
Add JSON Schema version identification field

Leinnan committed:
Clean up JSON schema test messages

Leinnan committed:
Add InvalidJsonSchema error and helper impl for TypeRegistration

Leinnan committed:
Use Cow instead of String in JSON schema types

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Test without documentation fix, typo fixes

Leinnan committed:
Clippy fixes

Leinnan committed:
Format code

Leinnan committed:
Rename reflect_types to reflect_type_data, update field doc

Leinnan commented:
I also was wondering about adding dev dependency: https://github.com/GREsau/schemars 
It would help out with validation whether generated schemas are valid and test out more scenarios that should pass/fail for given type. Is that a justified use case?

Leinnan committed:
values bound fix

splo commented:
Here is a gist of the changes of schema of `Transform` (as an example) between `main` and after this PR at this time: <https://gist.github.com/splo/76ca6c70dd1267f1222c8425262bcdbd/revisions>.

-   `"$schema": "https://json-schema.org/draft/2020-12/schema"` is added for every single type. Is this wanted? It is a bit noisy. Ideally it should be only present once at the top of the document. More on this below.
- `default` and `description` are added and I can't stress enough how this is awesome! 😁
- `reflectTypes` renamed into `reflectTypeData`. I don't care but if I must nitpick, I'd say the plural of `types` implied an array while the new `Data` suffix is quite unexpressive. Unless there are plans to change from an array to an struct?
- The most important change is that there were references (`"translation": {"type": {"$ref": "#/$defs/glam::Vec3"}`) and now the definition are fully expanded: the whole `Vec3` is repeated 3 times (for `rotation`, `scale` and `translation`). Is this what we want? It makes it way easier for tools to reflect on data, but it makes it a lot heavier.
- I don't mention them but there are other awesome changes, no comment about them.

So this pull request improves things a lot.

But ideally, instead of returning a struct of JSON schemas such as this:
```json
{
  "id": 0,
  "jsonrpc": "2.0",
  "result": {
    "glam::Vec3": {/* Full JSON Schema of Vec3 */},
    "bevy_transform::components::global_transform::GlobalTransform": {/* Full JSON Schema of GlobalTransform */},
    "bevy_transform::components::transform::Transform": {/* Full JSON Schema of Transform */}
  }
}
```
It should, imho, return a single JSON schema with definitions and references:
```json
{
  "id": 0,
  "jsonrpc": "2.0",
  "result": {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$defs": {
      "glam::Vec3": {/* Full JSON Schema of Vec3 */},
      "bevy_transform::components::global_transform::GlobalTransform": {/* JSON Schema of GlobalTransform with refs to Vec3 */},
      "bevy_transform::components::transform::Transform": {/* JSON Schema of Transform with refs to Vec3 */}
     }
  }
}
```
I hope it's clear enough.

Anyway if this is not the place, or too complicated to do in this PR, feel free to ignore this. 🙂

splo commented:
By the way, there could be an argument to the RPC method to get schema to configure whether the result should have definitions and references, vs it should expand every types. Best of both world but even more complex to implement! 😄

Leinnan committed:
Update crates/bevy_remote/src/lib.rs

Good call

Co-authored-by: Sébastien Job <sebastien.job@gmail.com>

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan commented:
> reflectTypes renamed into reflectTypeData. I don't care but if I must nitpick, I'd say the plural of Types implied an array while the new Data suffix is quite unexpressive. Unless there are plans to change from an array to a struct?

@splo It was requested [here](https://github.com/bevyengine/bevy/pull/19524#discussion_r2133973694) by @MrGVSV .

> default and description are added and I can't stress enough how this is awesome! 😁

> I don't mention them but there are other awesome changes, no comment about them.

![oh stop it you](https://images-cdn.9gag.com/photo/4924097_700b.jpg)

> "$schema": "https://json-schema.org/draft/2020-12/schema" is added for every single type. Is this wanted? It is a bit noisy. Ideally it should be only present once at the top of the document. More on this below.

> The most important change is that there were references ("translation": {"type": {"$ref": "#/$defs/glam::Vec3"}) and now the definition are fully expanded: the whole Vec3 is repeated 3 times (for rotation, scale and translation). Is this what we want? It makes it way easier for tools to reflect on data, but it makes it a lot heavier.

> By the way, there could be an argument to the RPC method to get schema to configure whether the result should have definitions and references, vs it should expand every types. Best of both world but even more complex to implement! 😄 

I think I agree with you that the best option would be to add support for `$defs`. I did it already in one of my previous iterations on this topic, but here didn't add at first for the sake of simplicity. I agree that it would be nice to have that configurable, especially for exporting schemas for editors, because the easiest way to use them is to have all definitions recursively for each asset type.

That also means that the response for RPC endpoint would change a bit, but I think it is justified change.

Leinnan committed:
Cleanup and refactor

Leinnan committed:
Start populating schema definitons field

Leinnan committed:
Schema definitions generate better

Leinnan committed:
Cleanup

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
schema IDs using URN format, further improvements, tests for export
endpoint

Leinnan committed:
Format fix

Leinnan committed:
Docs fix

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Use oneOf for Option types, use jsonschema for testing schemas

ChristopherBiscardi commented:
**My biggest concern is that I don't see any versioning on the endpoint, so a consumer won't know which api response type to use between 0.16 and 0.17.**

I would highly prefer to not force Blender users to pick a Bevy version (especially since they often *won't know*), so having that endpoint versioning is important. If there is a version or other field that is available that I missed please let me know.

---

> Currently implemented JSON Schema export is often producing invalid schemas for types resulting in that json that would be a valid value for schema generated for type cannot be deserialized to that type by Bevy deserializer.

It would be nice to have examples of what invalid schemas are currently being produced, so that we have a group understanding of what this PR is trying to fix. Right now it feels like there are superficial breakages like: `"type": {"$ref": "#/$defs/f32"} would be {"type": "number", "shortPath": "f32", "typePath": "f32"}` that don't have much of benefit to an API consumer.

---

As someone who has already used the old endpoint to implement a user-driven interface for constructing Component values, its not clear to me that this is an improvement over the old api. The adherence to jsonschema seems driven by VSCode's needs specifically and without doing a full re-implementation I won't be able to find the edge cases.

The test cases also have functional changes which make it harder to review. ex: was previously an `f32` is now an `integer`. An integer with min/max values is fine but also seems to imply that `NonZeroI8` is unrepresentable.

![image](https://github.com/user-attachments/assets/c9222e24-1ac9-42f5-af45-23d501b556d1)

---

Including Default values is useful and probably could've been a separate, smaller PR from this one. Default values aren't the only ones that people end up wanting though. Components can have multiple "presets" and I've also included the ability to define arbitrary name-value pairs for any Component type in Skein, which is likely also going to be useful for the upstream endpoint long-term. Ex:

```rust
App::new()
    .insert_skein_preset("Red", PointLight {
        color: Color::RED,
        intensity: 1_000_000.0,
        range: 20.0,
        radius: 0.0,
        shadows_enabled: false,
        affects_lightmapped_mesh_diffuse: true,
        shadow_depth_bias: 0.08,
        shadow_normal_bias: 0.6,
        shadow_map_near_z: 0.1,
    })
```

---

Overall I'm unconvinced that adding jsonschema support is a net benefit but there are nice pieces that could be split out (like Default values). I won't block the PR since worst-case scenario Skein can copy/paste the old endpoint behavior for 0.17 and will try to find some time to do an implementation to give some feedback.

Leinnan committed:
Remove id field

Leinnan committed:
Basic support for NonZero types with `not` field

Leinnan committed:
Fixes for NonZero types, store typepath for primitive types

Leinnan committed:
Restore a as a f32 field, add separate unsigned field in test struct

splo commented:
> **My biggest concern is that I don't see any versioning on the endpoint, so a consumer won't know which api response type to use between 0.16 and 0.17.**

There is the `rpc.discover` method that returns the Bevy version:

```json
{
  "jsonrpc": "2.0",
  "id": 0,
  "result": {
    "info": {
      "title": "Bevy Remote Protocol",
      "version": "0.16.0"
    },
    "methods": [],
    "openrpc": "1.3.2",
    "servers": []
  }
}
```

Leinnan commented:
@ChristopherBiscardi I don't get the argument about the nonzero types- were they supported before? But I've added some support for them now ;) 

I just also added `typePath` field for those primitve types. It should make easier to map them into types in Rust.

> My biggest concern is that I don't see any versioning on the endpoint, so a consumer won't know which api response type to use between 0.16 and 0.17.

Old version was returning hashmap of strings and schemas. New returns "empty" top level schema with schemas stored in `$defs` field.

BTW This improved compatibility with JSON Schema doesn't mean that you could use libraries like [this one](https://python-jsonschema.readthedocs.io/en/stable/) for your plugin?

Leinnan committed:
Add helper trait for registering schema types

Leinnan commented:
@splo 
Wrong types that should be Strings- if I find some time for it- sure, if that doesn't make into that PR I would suggest creating issues once this once is merged and work on them then :) 

> Schemas that have references ($ref) also get other fields (like type).

It should be better now, if not, please send me some examples, I will improve based on them.

> Option type names might not be unique.

Is that still the case? 

URN is gone.

> Primitive types have disappeared. It's cool but I wonder if this will bite us later. As for myself, I didn't mind having those f32, u8, etc: it allowed to know the actual types Bevy use. But yes it meant another redirection (every primitive fields had to reference the primitives).

For primitive types now I restored typePath field. So at least you have path to the type that is below it.

> Hash maps don't define their values properly.

It should be better now, but I am open to suggestion how to improve it further.

About renaming shortPath to title- I would move it into separate thing. It is already way to big.
Same for the readOnly field IMHO.

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Cleanup, support for Cow<str>

Leinnan committed:
Missing definitions handling

splo commented:
> @splo Wrong types that should be Strings- if I find some time for it- sure, if that doesn't make into that PR I would suggest creating issues once this once is merged and work on them then :)

Ok! That's a bigger issue we'll tackle in another PR.
Keep in mind these should not all be strings. Like `Entity` that's defined as an object but actually is a number.

> > Schemas that have references ($ref) also get other fields (like type).
> 
> It should be better now, if not, please send me some examples, I will improve based on them.

There's improvement but I still see, for instance `Transform` (with other fields redacted):
```json
{
  "properties": {
    "rotation": {
      "$ref": "#/$defs/glam-Quat",
      "description": "Rotation of the entity. See the [`3d_rotation`] example for usage. [`3d_rotation`]: https://github.com/bevyengine/bevy/blob/latest/examples/transforms/3d_rotation.rs",
      "type": "object"
    },
    "scale": {
      "$ref": "#/$defs/glam-Vec3",
      "description": "Scale of the entity. See the [`scale`] example for usage. [`scale`]: https://github.com/bevyengine/bevy/blob/latest/examples/transforms/scale.rs",
      "type": "object"
    },
    "translation": {
      "$ref": "#/$defs/glam-Vec3",
      "description": "Position of the entity. In 2d, the last value of the `Vec3` is used for z-ordering. See the [`translations`] example for usage. [`translations`]: https://github.com/bevyengine/bevy/blob/latest/examples/transforms/translation.rs",
      "type": "object"
    }
  }
}
```
It's not very important IMHO.

> > Option type names might not be unique.
> 
> Is that still the case?

It's still the case for the key name. Here is `(u8, u8)`:
```json
{
  "$defs": {
    "(u8, u8)": {
      "crateName": "core",
      "kind": "Optional",
      "modulePath": "core::option",
      "oneOf": [
        {
          "type": "null"
        },
        {
          "$ref": "#/$defs/(u8, u8)",
          "type": "object"
        }
      ],
      "shortPath": "Option<(u8, u8)>",
      "typePath": "core::option::Option<(u8, u8)>"
    }
  }
}
```
Key is `"(u8, u8)"` but should be `core-option-Option<(u8, u8)>`. It's even worse that it recursively refers to itself 😆. I guess the real `(u8, u8)` was defined but overwritten by the optional one.

❓ By the way, what is the reason why you changed key names replacing `::` with `-`? Did colons prove problematic?

> URN is gone.

👍

> For primitive types now I restored typePath field. So at least you have path to the type that is below it.

👍

> > Hash maps don't define their values properly.
> 
> It should be better now, but I am open to suggestion how to improve it further.

I didn't notice anything new there. Still `keyType` and `valueType`. It's ok, that can change in the future.

> About renaming shortPath to title- I would move it into separate thing. It is already way to big. Same for the readOnly field IMHO.

Yes I agree.

Leinnan commented:
@splo Colons were problematic, the rest of the issues I will try to address later this weekend.

Leinnan committed:
Reworked building type dependencies list

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan commented:
> Key is "(u8, u8)" but should be core-option-Option<(u8, u8)>. It's even worse that it recursively refers to itself 😆. I guess the real (u8, u8) was defined but overwritten by the optional one.

It should be better now

splo commented:
No, for instance:
```json
    "alloc-vec-Vec<alloc-string-String>": {
      "crateName": "core",
      "kind": "Optional",
      "modulePath": "core::option",
      "oneOf": [
        {
          "type": "null"
        },
        {
          "$ref": "#/$defs/alloc-vec-Vec<alloc-string-String>"
        }
      ],
      "shortPath": "Option<Vec<String>>",
      "typePath": "core::option::Option<alloc::vec::Vec<alloc::string::String>>"
    },
```

Leinnan committed:
Move more logic into trait for TypeRegistry, better defs id generation

Leinnan committed:
More logic in trait, start removing old code

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Ref type means no schema type

Leinnan commented:
@splo I cannot stress enough how useful your testing is! Thanks, I was using your way of testing out to detect a lot of the issues, now it should be better. I already started removing some of the older code and moving more and more into trait where TypeRegistry can do more heavy lifting :)

ChristopherBiscardi commented:
> @ChristopherBiscardi I don't get the argument about the nonzero types- were they supported before? But I've added some support for them now ;)
> 
> I just also added `typePath` field for those primitve types. It should make easier to map them into types in Rust.

Great, thanks! having the actual type names is useful for communicating what a user is actually supposed to be targeting when building a Component value. Especially since the type path is used today in the old endpoint.

> > My biggest concern is that I don't see any versioning on the endpoint, so a consumer won't know which api response type to use between 0.16 and 0.17.
> 
> Old version was returning hashmap of strings and schemas. New returns "empty" top level schema with schemas stored in `$defs` field.

"the response data is different" isn't a versioning scheme... This means clients will have to guess at compatibility as the endpoints evolve. The lowest effort lift to fix it would likely be to include the bevy_remote version from the Cargo.toml in the response so that clients can determine which bevy_remote version they're talking to, and thus which endpoint they're using. It wouldn't be perfect, but it would be something.

Also look at it from the perspective of existing clients. This is a breaking API change on the same url so 0.16 compatible clients will stop working without a good error message even though the old endpoint would still have worked, afaict.

> BTW This improved compatibility with JSON Schema doesn't mean that you could use libraries like [this one](https://python-jsonschema.readthedocs.io/en/stable/) for your plugin?

I don't see how this would be useful tbh, maybe you could explain the usecase you think this would solve?

The source of truth for a valid value is not actually the jsonschema validation suggested here but whether the reflection infrastructure accepts the value or not. Skein doesn't let users construct invalid Component values and that's an extension concern not an end-user concern. Blender users don't actually have control over the value structure or the types of values (and a 3d artist isn't going to solve "the jsonschema failed to validate" anyway); the extension defines the interface for them, so if a Blender user has an invalid value as determined by that library there's nothing they can do because the extension would be constructing it incorrectly for them.

Leinnan committed:
Remove some of the feature specific types, simplify the code

Leinnan commented:
> "the response data is different" isn't a versioning scheme... This means clients will have to guess at compatibility as the endpoints evolve. The lowest effort lift to fix it would likely be to include the bevy_remote version from the Cargo.toml in the response so that clients can determine which bevy_remote version they're talking to, and thus which endpoint they're using. It wouldn't be perfect, but it would be something.

So what you suggesting is adding a field to json schema with bevy_remote version? Or I did get that wrong?

If so, maybe? In this PR schema as top element contains field "$schema" with value "https://json-schema.org/draft/2020-12/schema", that describes what format should be expected. Maybe we could add information about used bevy_remote version, how you would see that?

> Also look at it from the perspective of existing clients. This is a breaking API change on the same url so 0.16 compatible clients will stop working without a good error message even though the old endpoint would still have worked, afaict.

I totally get that, this is why I really wanted feedback from you in the first place. Issue I had with the existing format is that it is "kinda" like json schema, but it is not a valid one. It makes it not so easy to use it with libraries for json schema and even reason if something is a valid schema, or a bug.

> I don't see how this would be useful tbh, maybe you could explain the usecase you think this would solve?
> The source of truth for a valid value is not actually the jsonschema validation suggested here but whether the reflection infrastructure accepts the value or not. Skein doesn't let users construct invalid Component values and that's an extension concern not an end-user concern. Blender users don't actually have control over the value structure or the types of values (and a 3d artist isn't going to solve "the jsonschema failed to validate" anyway); the extension defines the interface for them, so if a Blender user has an invalid value as determined by that library there's nothing they can do because the extension would be constructing it incorrectly for them.

Using JSON schema should allow you not only to detect if given value is valid, but a lot of libraries also offer human readable errors what went wrong when something does not match schema. You could have field and types descriptions, as well as ranges, etc.

I think having option to know that field in component should be a number that is not negative is valuable information, same for programmer and artist

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Register more glam types as arrays

Leinnan committed:
Update Cargo.toml

Leinnan committed:
Fix compilation, add more glam types

Leinnan commented:
@ChristopherBiscardi Fixed just now 🙈

ChristopherBiscardi commented:
Are defs and type paths supposed to be different? ex: `bevy_animation-ActiveAnimation` vs `bevy_animation::ActiveAnimation`

![screenshot-2025-07-07-at-09 13 28@2x](https://github.com/user-attachments/assets/b918c663-ad6d-4e03-9924-a5cbb1f87505)

![screenshot-2025-07-07-at-09 15 03@2x](https://github.com/user-attachments/assets/ac836879-8583-47c6-82ff-db647ab8ecb8)

Leinnan commented:
> Are defs and type paths supposed to be different?

Yes, to large extend def is type path with potentially problematic characters replaces/removed. That way the information about the type path is preserved and def ids can be resolved by json schema libraries.

ChristopherBiscardi commented:
> > Are defs and type paths supposed to be different?
> 
> Yes, to large extend def is type path with potentially problematic characters replaces/removed. That way the information about the type path is preserved and def ids can be resolved by json schema libraries.

ok so clients are expected to re-implement this slugging algorithm if they want to turn a type path into a key for the object. the type_paths are required data for constructing a valid reflected component value afaik and are how skein stored data is keyed currently. I guess I could iterate over all of the key/value pairs and .find the typePath matches instead.

I'm not entirely sure why :: and () are problematic for a string key value. What are we protecting against?

ChristopherBiscardi commented:
> Key name only becomes description when we are forcing struct to be array and field does not contain any description by itself.

ok so this field can not be used reliably for anything but displaying additional context in a popover/hover tooltip, etc. Clients effectively lose the x/y/z field names in the ui for all glam types then since the description could change at any time in the future. I'll likely end up hardcoding it for the glam types and not having it for user-defined types that are forced to arrays.

Leinnan committed:
TypeReferenceId RFC 3986 support improvements

Leinnan committed:
export schema now stores possible values to serialize in top 'one_of' field, support for Unit like structs serialization"

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan commented:
@ChristopherBiscardi I've updated top level schema generation a bit. Now it generates a schema you can use to validate against values that in the format used by ReflectSerializer. You can take look at the new `reflection_serialization_tests` test to check it out.

Leinnan committed:
Remove eprint

Leinnan committed:
Merge branch 'main' into proper-json-schema

ChristopherBiscardi commented:
> > I'm not entirely sure why :: and () are problematic for a string key value. What are we protecting against?
> 
> AFAIK they are just kinda reserved characters in RFC 3986, which is the standard that definitions ids should follow. https://www.tech-invite.com/y35/tinv-ietf-rfc-3986.html#e-2-2 But how is that id constructed on this crate side could be improved, if you have ideas for that. Typical type path can be a invalid value according to RFC 3986, so it cannot always be exactly that.

Why is this PR trying to follow an RFC that bevy itself doesn't follow? I don't see the reason for avoiding type paths in the first place. If the only problem is that it doesn't follow a spec that has no impact on Bevy usage then I propose we drop adherence to that spec.

Leinnan committed:
Make one_of field in top schema optional, disabled by default.

Leinnan commented:
> What is the intended use case of the new top-level `oneOf` with references to all types?

It allows to use the schema to validate against values to ensure that format is the same as would ReflectSerializer format the same data. But I totally get your issue. I made it into optional thing, that needs to be specificaly requested.

splo commented:
If I read it correctly, [RFC 3986 - 2.2 Reserved Characters](https://www.rfc-editor.org/rfc/rfc3986.html#section-2.2) suggests using percent-encoding in order to construct correct URI references. I feel that would be a good solution.

splo commented:
> I was trying out this as solution, but https://github.com/Stranger6667/jsonschema was claiming that it is not a valid schema, so I did revert this change to the one currently existing in the PR. But I am open to improve it.

After testing with this library, the following is a valid schema:
```json
{
  "$ref": "#/$defs/%28foo%3A%3Abar%2C%20baz%3A%3Aqux%29",
  "$defs": {
    "(foo::bar, baz::qux)": {
      "type": "number"
    }
  }
}
```

That is, the percent encoding part should only be in the reference, not at the definition key. The only thing to care is not to convert spaces with `+` but with `%20`, meaning valid percent encoding and not URL encoding.

Then we could satisfy @ChristopherBiscardi: having valid references and keep type paths as keys.

Leinnan committed:
Allow providing CustomInternalSchemaData instead of only adding data for
forcing as array.

Leinnan committed:
Simplify encoding of the def ids

Leinnan committed:
Merge branch 'main' into proper-json-schema

splo commented on crates/bevy_remote/src/schemas/reflect_info.rs:
```diff
@@ -0,0 +1,2332 @@
+#[derive(Debug, Clone, PartialEq, Default, Reflect, Hash, Eq, Ord, PartialOrd)]
+pub struct TypeReferencePath {
+    /// The location of the reference in the JSON schema.
+    pub localization: ReferenceLocation,
```
`localization` or `location`?
```suggestion
    pub location: ReferenceLocation,
```

Leinnan replied:
Good call, updated

Leinnan committed:
Update TypeReferencePath

Leinnan committed:
Merge branch 'main' into proper-json-schema

splo commented:
When running the example server, it panics with the following error:
```
thread 'main' panicked at crates/bevy_remote/src/schemas/mod.rs:151:14:
SHOULD NOT HAPPENED
stack backtrace:
   5: <bevy_reflect::type_registry::TypeRegistry as bevy_remote::schemas::RegisterReflectJsonSchemas>::register_data_type_by_value
             at ./crates/bevy_remote/src/schemas/mod.rs:150:9
   6: <bevy_app::app::App as bevy_remote::schemas::RegisterReflectJsonSchemas>::register_data_type_by_value
             at ./crates/bevy_remote/src/schemas/mod.rs:181:9
   7: bevy_remote::schemas::RegisterReflectJsonSchemas::register_force_as_array
             at ./crates/bevy_remote/src/schemas/mod.rs:122:9
   8: bevy_remote::schemas::RegisterReflectJsonSchemas::register_schema_base_types
             at ./crates/bevy_remote/src/schemas/mod.rs:62:13
   9: <bevy_remote::RemotePlugin as bevy_app::plugin::Plugin>::build
             at ./crates/bevy_remote/src/lib.rs:658:9
  10: bevy_app::app::App::add_boxed_plugin::{{closure}}
             at ./crates/bevy_app/src/app.rs:484:37
```
It seems it fails calling this at `RegisterReflectJsonSchemas::register_schema_base_types`:
```rust
            self.register_force_as_array::<bevy_math::I8Vec2>();
```

Leinnan committed:
Introduce fields: title, rust_fields_info, replace pattern_properties
with pattern and property_names

splo commented:
Before this PR, the registry query lied about some of its types. For instance, in my VS Code extension, I have code like this:

```ts
if (typePath === 'glam::Vec3`) {
  // Ignore that this pretends to be an object with x, y, z properties.
  // Instead treat it as an array with 3 elements.
}
```

I reported this issue earlier in this PR and I thought we concluded that we'd tackle it later. Indeed, the issue is not only with `bevy_math` types but (I think) for every type that has a custom serde implementation (for instance `core::time::Duration`).

The current state of this PR tries to manually fix the bevy_math types by hard-coding them while I hoped the fix would be more general and cover all such cases. I understand this is a tricky issue.

So first question: is it ok to have a temporary and incomplete workaround in this PR?

If so and we naively fix `bevy_math` types by reporting their serialized type, then indeed we lose some of the underlying internal Rust type information. We still have the most important though: the type path.

Second question: is it that important? What is the use case for reporting the internal Rust type information in the registry query?

I feel that is an actual controversial change. I think most BRP users only need to know the actual types they are supposed to work with. Is it about code generation? Like read type info from BRP and generate Rust code that should be compiled with Bevy types?

If we decide to somehow report the internal Rust type information, then yes, I guess we could include that in a custom field.

In the current format, is it possible to map the internal Rust field to its associated JSON Schema field? I mean for `DAffine3`, JSON schema value `[2]` maps to Rust value `matrix3.x_axis.z`. Is this what you're doing with the `title` field? If so, I'm not sure `title` is supposed to be used algorithmically, as it is more of a human-readable label. Then what alternatives do we have? It is a bit clunky.

---

Anyway, how about concentrating all custom fields (such as `typePath`, `shortPath`, `kind`, etc) that are not part of the JSON Schema standard under a single field? Something like `rustType` or `bevyType` or just `bevy`?

```json
"glam::Vec3": {
  "type": "array",
  "maxItems": 3,
  "minItems": 3,
  "prefixItems": [
    {
      "title": "x",
      "type": "number",
      "bevyType": {
        "kind": "Value",
        "typePath": "f32"
      }
    },
    {
      "title": "y",
      "type": "number",
      "bevyType": {
        "kind": "Value",
        "typePath": "f32"
      }
    },
    {
      "title": "z",
      "type": "number",
      "bevyType": {
        "kind": "Value",
        "typePath": "f32"
      }
    }
  ],
  "bevyType": {
    "crateName": "glam",
    "kind": "Struct",
    "modulePath": "glam",
    "reflectTypeData": ["Default", "Serialize", "Deserialize"],
    "shortPath": "Vec3",
    "typePath": "glam::Vec3",
    "fields": {
      "x": {
        "kind": "Value",
        "typePath": "f32"
      },
      "y": {
        "kind": "Value",
        "typePath": "f32"
      },
      "z": {
        "kind": "Value",
        "typePath": "f32"
      }
    }
  }
}
```

It would isolate non-standard fields and make it easier to parse the JSON Schema. It would also allow us to add more custom fields in the future without cluttering the main schema structure.

Leinnan committed:
Remove ReflectJsonSchema, functionality is provided by CustomInternalSchemaData

Leinnan commented:
@mweatherley @MrGVSV How do you feel about the current state of the PR?

Leinnan committed:
Update docs

splo commented:
> For this format I was thinking about what format for title would be useful to cover cases like the one here: https://github.com/rust-adventure/skein/blob/main/extension/object_to_form.py

I need more context information to understand exactly why the Rust internal structure is needed in this Python code. But fine, if there's demand for this feature, we can add it.

> I really like the idea of putting bevy specific fields into one field, but I don't know if that should be part of this PR tbh.

It could be in another PR of course. I just fear a PR with only cosmetic schema-breaking changes would be de-prioritized to hell. Like "too late, we can't change anything anymore, the API is frozen forever". I've seen that in a professional context so many times.

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Option for specifying default value presence

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Improve naming, reduce SchemaNumber type size

Leinnan commented on crates/bevy_remote/src/schemas/reflect_info.rs:
```diff
@@ -0,0 +1,2474 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ReferenceLocation::Definitions => write!(f, "#/$defs/"),
+            ReferenceLocation::Components => write!(f, "#/components/"),
```
Agree, but I that could be a separate improvement.

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan commented:
@splo thanks for the feedback! About the last two pieces- I think maybe it would make sense to move the schema stuff to separate crate **inside** bevy repo since it depends so much on bevy reflection.

Leinnan commented:
@MrGVSV I've tried to update first message in PR as well, do you have any opinions on the PR?

splo commented:
> @splo thanks for the feedback! About the last two pieces- I think maybe it would make sense to move the schema stuff to separate crate **inside** bevy repo since it depends so much on bevy reflection.

I understand much of the code rely on `bevy_reflect` but I was commenting on 2 different matter:
- The long reflect info file where I was asking if you (or anyone) knew any lib that could simplify the code. I don't know how to express, it seems very "type info" stuff that could live in `bevy_reflect` or even maybe some non-reflect part could be extracted out?
- The JSON Schema file comment is not about `bevy_reflect` stuff. The point is to more easily understand and maintain where Bevy is actually using the JSON Schema standards, and where it diverges to add its extensions. Also to reduce responsibilities and number of lines of code Bevy maintainers will have to handle.

splo reviewed (changes requested):
A migration guide is needed.

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Migration guide

Leinnan committed:
Minor fix

Leinnan committed:
Update json_schema_restructure.md

splo commented on release-content/migration-guides/json_schema_restructure.md:
```diff
@@ -0,0 +1,15 @@
+
+Types identifiers are stored in `definitions` as strings, but when they are referenced from other types by `$ref` key they are percent encoded to take into account fact that they have to be valid URI and don't use reserved characters. That means that the type `glam::Vec3` is stored by `glam::Vec3` key in `definitions` and is referenced as `glam%3A%3AVec3` in other types.
+
+There is added support for custom schemas for reflected types. One example of it are various glam types, which in Rust are represented as structs, but they are serialized into arrays of numbers. Since JSON Schema is a format for describing the structure of stored data, it now describes them as arrays of numbers. It is possible to get the information about the type by inspecting the additional `ru ...
```
I suggest adding a JSON sample of before and after, even though it is a bit verbose. Here is one reduced to `bevy_input::mouse::AccumulatedMouseMotion` and `glam::Vec2`:

Before:
```json
{
  "jsonrpc": "2.0",
  "id": 0,
  "result": {
    "bevy_input::mouse::AccumulatedMouseMotion": {
      "additionalProperties": false,
      "crateName": "bevy_input",
      "kind": "Struct",
      "modulePath": "bevy_input::mouse",
      "properties": {
        "delta": { "type": { "$ref": "#/$defs/glam::Vec2" } }
      },
      "reflectTypes": ["Resource", "Default", "Serialize", "Deserialize"],
      "required": ["delta"],
      "shortPath": "AccumulatedMouseMotion",
      "type": "object",
      "typePath": "bevy_input::mouse::AccumulatedMouseMotion"
    },
    "glam::Vec2": {
      "additionalProperties": false,
      "crateName": "glam",
      "kind": "Struct",
      "modulePath": "glam",
      "properties": {
        "x": { "type": { "$ref": "#/$defs/f32" } },
        "y": { "type": { "$ref": "#/$defs/f32" } }
      },
      "reflectTypes": ["Default", "Serialize", "Deserialize"],
      "required": ["x", "y"],
      "shortPath": "Vec2",
      "type": "object",
      "typePath": "glam::Vec2"
    }
  }
}
```

After:
```json
{
  "jsonrpc": "2.0",
  "id": 0,
  "result": {
    "$defs": {
      "bevy_input::mouse::AccumulatedMouseMotion": {
        "additionalProperties": false,
        "crateName": "bevy_input",
        "default": {
          "delta": [0.0, 0.0]
        },
        "description": "Tracks how much the mouse has moved every frame. This resource is reset to zero every frame. This resource sums the total [`MouseMotion`] events received this frame.",
        "kind": "Struct",
        "modulePath": "bevy_input::mouse",
        "properties": {
          "delta": {
            "$ref": "#/$defs/glam%3A%3AVec2",
            "description": "The change in mouse position.",
            "kind": "Struct",
            "title": "delta",
            "typePath": "glam::Vec2"
          }
        },
        "reflectTypeData": ["Default", "Serialize", "Deserialize", "Resource"],
        "required": ["delta"],
        "shortPath": "AccumulatedMouseMotion",
        "type": "object",
        "typePath": "bevy_input::mouse::AccumulatedMouseMotion"
      }
    },
    "glam::Vec2": {
      "crateName": "glam",
      "default": [0.0, 0.0],
      "kind": "Struct",
      "maxItems": 2,
      "minItems": 2,
      "modulePath": "glam",
      "prefixItems": [
        { "kind": "Value", "title": "x", "type": "number", "typePath": "f32" },
        { "kind": "Value", "title": "y", "type": "number", "typePath": "f32" }
      ],
      "reflectTypeData": ["Default", "Serialize", "Deserialize"],
      "rustFieldsInfo": {
        "x": { "kind": "Value", "title": "x", "type": "number", "typePath": "f32" },
        "y": { "kind": "Value", "title": "y", "type": "number", "typePath": "f32" }
      },
      "shortPath": "Vec2",
      "type": "array",
      "typePath": "glam::Vec2"
    }
  }
}
```

Leinnan committed:
Update release-content/migration-guides/json_schema_restructure.md

Co-authored-by: Sébastien Job <sebastien.job@gmail.com>

Leinnan committed:
Update release-content/migration-guides/json_schema_restructure.md

Co-authored-by: Sébastien Job <sebastien.job@gmail.com>

Leinnan committed:
Add example, fix link

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Update json_schema_restructure.md

splo reviewed (approved):
Wonderful work, congratulations!

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
combine if statements

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Update reflect_info.rs

Leinnan committed:
Update reflect_info.rs

Leinnan committed:
Sort get_registered_reflect_types response

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Update reflect_info.rs

Leinnan commented:
@ChristopherBiscardi thanks for letting know about potential improvements, I will try to improve it. 

There is one thing I disagree with you totally, which is that this change is mostly due to adhere to a spec. I blame it on my poor communication skills. 😅  This PR is result of me trying to solve two things:
- Bevy has initial support for OpenRPC. It would be great to improve that to have support for information about arguments and response schemas. To have that we still would need a valid json schema support in engine, like the one introduced in this PR. This PR opens a way to solve #16744 
- I wanted to have a better DX for editing bevy assets from code editors like JetBrains IDEs ([docs](https://www.jetbrains.com/help/rust/json.html#ws_json_schema_add_custom), VSCode or Zed. 

Making schemas in Bevy adhere to a spec solves that issues.

> Option is now treated as a required field with possible values of T or null
> This option change points quite clearly to the validation enabled by jsonschema not being "what is necessary to reflect a value" but rather a completely separate target all its own. So jsonschema is not actually validating whether a value will correctly reflect or not, rendering the validation enabled by adhering to the spec somewhat deficient.

Could you give me a example when that will not validate if some value is a valid one? 
To give example of my proposed implementation:
- if I have a `Option<String>` type. JSON Schema for it would be something like: 
```json 
{
  "oneOf": [
    {
      "type": "null"
    },
    {
      "type": "string"
    }
  ]
}
```
For that schema `"OP"` `"T"`, `null` would be valid values, but `15`, `{"test":"D"}` or `[""]` would be not. That is consistent with how serde serialize values as well, `Option` with `None` value would be serialized into `null`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=fb9ea4dd50f27d8805d8d5691b61994a

Leinnan committed:
Improve type names

Leinnan commented:
> The declaration that an optional field is required is incorrect from the perspective of reflection usage. Any field on a type with Default derived is a trivial counter-example where the key doesn't need to be present in json to serialize correctly.

@ChristopherBiscardi Okay, that is interesting- so the struct below should have no required fields then? 
```rust
#[derive(Reflect, Resource, Default, Deserialize, Serialize)]
#[reflect(Resource, Default, Serialize, Deserialize)]
struct Foo {
  a: f32,
  #[reflect(@10..=15i16)]
  b: Option<i16>,
}
```

And this one should have only the last field as required?
```rust
#[derive(Reflect, Resource, Deserialize, Serialize)]
 #[reflect(Resource, Serialize, Deserialize)]
 struct TestString(pub String);
#[derive(Reflect, Resource, Deserialize, Serialize)]
#[reflect(Resource, Serialize, Deserialize)]
struct Foo {
  a: f32,
  #[reflect(@10..=15i16)]
  b: Option<i16>,
  required: TestString,
}
```

Leinnan committed:
Fix

ChristopherBiscardi commented:
> If I understand your approach here you would like that this one:

I feel like stating this as a personal preference is taking away from the technical point I'm making.

This PR does not have a coherent definition of what "required" means as far as I can tell. It clearly does not mean the field key is required, as shown in previous comments. Thus the validation feature that is built on top of this information seems broken on day 1, since it doesn't try to target what is actually accepted by the reflection infrastructure but introduces a different target.

For example, the validation using required fields here means that the author of the json content in your use case can't defer a value to a Default implementation by leaving a field out. They *must* specify a value as that is what has been dictated by the validation information.

> Should not have any fields required then?

No fields are required to exist in the JSON to deserialize a valid value, so if that's what the PR wants to define as "required" then yes, none of these fields are required.

> Why in this case a is required and the b is not?

Because that is the concrete way the reflection infrastructure works with the traits that have been implemented on these types. A value can be deserialized in that example if `a` exists and `b` does not.

> When you mention requiring Serialize/Deserialize traits 

I'm not entirely sure what you're asking here as I didn't mention requiring Serialize/Deserialize traits. In that paragraph I am talking about the code that you posted specifically and the behavior described is contingent on the additional Serialize/Deserialize traits that were used. If those traits aren't implemented, the behavior changes.

Leinnan committed:
Optional fields in struct are no longer required

Leinnan commented:
I updated PR a bit. I got one thing about serde json wrong- I was thinking that if you have field with type `Option<T>` it needs to store either `null` or `T` value and if you want to make the field itself optional you need to add `#[serde(default)]` there. But it looks like in case of optionals if there is no field it can deserialize it as `None`. So starting now, fields in structs that are optional are not stored as required.

With that update:

> This PR does not have a coherent definition of what "required" means as far as I can tell. It clearly does not mean the field key is required, as shown in previous comments. Thus the validation feature that is built on top of this information seems broken on day 1, since it doesn't try to target what is actually accepted by the reflection infrastructure but introduces a different target.

My goal for this PR is to provide a way to create json schemas for types that could validate content so if it passes json schema validation then it for sure is a valid value and if I try to later on parse it with serde it should not fail. Serde being important part here since current RPC implementation uses serde specificaly for handling messages serialization. 

@ChristopherBiscardi does that with my mistake fixed gives you a more coherent definition of the `required` fields and the PR as whole?

Leinnan committed:
Update json_schema.rs

ChristopherBiscardi commented:
here's an example: https://github.com/bevyengine/bevy/pull/20940

ChristopherBiscardi commented:
re, key points:

In general I don't think we should be codifying a requirement to use serde's traits in this validation schema. If we're going to upstream a jsonschema-based validator feature, it should validate against the Reflect/SerializationData implementation. bevy_reflect is capable of deserializing new structs and doesn't need to rely on serde's traits being implemented on every type.

ChristopherBiscardi commented:
> one of the things the code has to watch for is whether Serialize/Deserialize is derived on a Component because if it's only Reflect, then I can't spawn/insert I can only Mutate

Not sure where you got that info; here's [insert_reflect](https://docs.rs/bevy/0.16.1/bevy/ecs/reflect/trait.ReflectCommandExt.html#tymethod.insert_reflect) for components.

Either way this thread isn't a feedback thread for the reflection features or brp in general. Perhaps you should post it in one of the relevant discord channels or a github discussion?

Leinnan committed:
Merge branch 'main' into proper-json-schema

natepiano commented:
> Either way this thread isn't a feedback thread for the reflection features or brp in general. Perhaps you should post it in one of the relevant discord channels or a github discussion?

Thanks for replying - reproduced it, proposed a fix, [made an issue](https://github.com/bevyengine/bevy/issues/20952)

Leinnan committed:
Support and tests for types without serde Serialize

Leinnan committed:
Update reflect_info.rs

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Update reflect_info.rs

Leinnan committed:
Fix test and typo

Leinnan committed:
Merge branch 'main' into proper-json-schema

Leinnan committed:
Update reflect_info.rs
